Various fixes to RPC `importaddresses`

Justaphf opened the pull request:
Noticed several issues with the RPC `importaddresses` call while trying to import about 30 watch-only addresses in one go.  This PR fixes these issues.

1. There was an incorrect max parameter size check.  This appears to be a copy-and-paste error from `importaddress`.  This prevented adding more than 4 addresses at at time (3 if the rescan|no-rescan option was explicitly specified)
2. Various corrections to the information in the help text
3. Incorrectly tries to set the label for all addresses imported to the value of the second parameter.  This is wrong because label(s) as input parameters are not supported by `importaddresses` and this caused every address imported to have the address in the 2nd parameter spot as it's label.

Justaphf committed:
Remove max param size check for RPC importaddresses

This appears to be a copy-and-paste error from importaddress, where the maximum number of params check makes sense.
This check prevented adding more than 4 addresses at a time, 3 if the rescan|no-rescan parameter was used

Justaphf committed:
Corrections to RPC importaddresses help text

Fixes parameter numbering, incorrect mention of address labels as input params, and incorrect RPC text calling importaddress (singular)

Justaphf committed:
Remove incorrect handling of address labels from RPC importaddresses

Labels for addresses are not part of the supported input parameters, so this erroneously set the label for all imported addresses to the value of the 2nd parameter when the rescan option or more more than 1 address was specified.

awemany commented on src/wallet/rpcdump.cpp:
```diff
@@ -363,16 +363,16 @@ UniValue importaddresses(const UniValue &params, bool fHelp)
             "\nArguments:\n"
             "1. \"rescan | no-rescan\" (string, optional default rescan) If \"no-rescan\", skip wallet rescan\n"
-            "1. \"address\"           (string, 0 or more) The address or hex-encoded P2SH script\n"
+            "2. \"address\"           (string, 0 or more) The address or hex-encoded P2SH script\n"
```
Shouldn't this be more like "2..." or so?

Justaphf replied:
I looked and the only other RPC call that takes a params[] style list like this (rather than a JSON Array) is `importprivatekeys` and it also follows this convention.  If we want to change we should update both places for consistency.

I'm not sure what is a standard convention for a help parameter that is a list of all subsequent parameters in the method call.  Using the ellipse seems very subtle, though I don't really have a better idea.

I also noticed that the way the method is written, adding an address is optional if you explicitly specify the resacn|no-rescan option, it would pass all the parameter checks, then fall through (and perform a rescan if that was the option) without attempting to add any addresses.  I think that maybe this should be fixed too to require at least one address, though I do sometimes find it useful to be able to initiate a rescan without needing to import anything or restart the node.

awemany replied:
> I'm not sure what is a standard convention for a help parameter that is a list of all subsequent parameters in the method call. Using the ellipse seems very subtle, though I don't really have a better idea.

Fair point. A "2." feels somewhat wrong to me but then this is a very minor thing. 

> I also noticed that the way the method is written, adding an address is optional if you explicitly specify the resacn|no-rescan option, it would pass all the parameter checks, then fall through (and perform a rescan if that was the option) without attempting to add any addresses. I think that maybe this should be fixed too to require at least one address, though I do sometimes find it useful to be able to initiate a rescan without needing to import anything or restart the node.

Hmm, maybe adding another line to the docs that you can do a rescan this way would be good to have then, to make this use-case explicit? I have no strong opinion either way...

Justaphf replied:
I added an extra line to the help text for rescan without importing.  It turns out that functionality is explicitly used in this manner for the QA tests.

awemany commented:
This looks good to me, though I only read the code and did not check correct functionality manually. 

In my opinion, it would be awesome if you could extend the tests of `importaddresses` in `qa/rpc-tests/wallet.py` to also test for the problems that you fixed here :)

Justaphf commented:
Putting into WIP until I update the QA tests as suggested by @sickpig

sickpig commented:
@Justaphf great! (the QA tests was suggested by @awemany :smile:, though)

Justaphf commented:
Oops, yes, fixed.  I looked at the name directly above my last comment which happened to be your approval rather than awemany's comment.

Justaphf committed:
Update wallet tests to verify fixes

Updates to tests made first to reproduce the bugs, which should be cleared up as additional commits are applied.

awemany reviewed (approved):
Great! I wonder whether there's an easy way to test the rescan-only `importaddresses` with an empty list.

Justaphf commented:
The rescan-only `importaddresses` was already already being used around lines 379 through 383 to in wallet.py tests (line numbers are from my commits).

1. Import some watch addresses which have funds, but don't rescan.  Verify the wallet balance hasn't changed.
2. Rescan without importing any new addresses.  Verify the wallet balances have updated for the new watch addresses that were previously added.
